// @graphql-tools/url-loader/index.esm.js
import { print } from 'graphql';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { isWebUri } from 'valid-url';
import { fetch } from 'cross-fetch';
import { introspectSchema, makeRemoteExecutableSchema } from '@graphql-tools/wrap';

class UrlLoader {
    loaderId() {
        return 'url';
    }
    async canLoad(pointer, _options) {
        return !!isWebUri(pointer);
    }
    canLoadSync(_pointer, _options) {
        return false;
    }
    async load(pointer, options) {
        let headers = {};
        let fetch$1 = fetch;
        let method = 'POST';
        if (options) {
            if (Array.isArray(options.headers)) {
                headers = options.headers.reduce((prev, v) => ({ ...prev, ...v }), {});
            }
            else if (typeof options.headers === 'object') {
                headers = options.headers;
            }
            if (options.customFetch) {
                if (typeof options.customFetch === 'string') {
                    const [moduleName, fetchFnName] = options.customFetch.split('#');
                    fetch$1 = await import(moduleName).then(module => (fetchFnName ? module[fetchFnName] : module));
                }
            }
            if (options.method) {
                method = options.method;
            }
        }
        const extraHeaders = {
            Accept: 'application/json',
            'Content-Type': 'application/json',
            ...headers,
        };
        const executor = async ({ document, variables }) => {
            const fetchResult = await fetch$1(pointer, {
                method,
                ...(method === 'POST'
                    ? {
                        body: JSON.stringify({ query: print(document), variables }),
                    }
                    : {}),
                headers: extraHeaders,
            });
            return fetchResult.json();
        };
        const clientSchema = await introspectSchema(executor);
        const remoteExecutableSchema = makeRemoteExecutableSchema({
            schema: printSchemaWithDirectives(clientSchema, options),
            executor,
        });
        return {
            location: pointer,
            schema: remoteExecutableSchema,
        };
    }
    loadSync() {
        throw new Error('Loader Url has no sync mode');
    }
}

export { UrlLoader };
