// @graphql-tools/url-loader/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopNamespace(e) {
    if (e && e.__esModule) { return e; } else {
        var n = {};
        if (e) {
            Object.keys(e).forEach(function (k) {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () {
                        return e[k];
                    }
                });
            });
        }
        n['default'] = e;
        return n;
    }
}

const graphql = require('graphql');
const utils = require('@graphql-tools/utils');
const validUrl = require('valid-url');
const crossFetch = require('cross-fetch');
const wrap = require('@graphql-tools/wrap');

class UrlLoader {
    loaderId() {
        return 'url';
    }
    async canLoad(pointer, _options) {
        return !!validUrl.isWebUri(pointer);
    }
    canLoadSync(_pointer, _options) {
        return false;
    }
    async load(pointer, options) {
        let headers = {};
        let fetch = crossFetch.fetch;
        let method = 'POST';
        if (options) {
            if (Array.isArray(options.headers)) {
                headers = options.headers.reduce((prev, v) => ({ ...prev, ...v }), {});
            }
            else if (typeof options.headers === 'object') {
                headers = options.headers;
            }
            if (options.customFetch) {
                if (typeof options.customFetch === 'string') {
                    const [moduleName, fetchFnName] = options.customFetch.split('#');
                    fetch = await new Promise(function (resolve) { resolve(_interopNamespace(require(moduleName))); }).then(module => (fetchFnName ? module[fetchFnName] : module));
                }
            }
            if (options.method) {
                method = options.method;
            }
        }
        const extraHeaders = {
            Accept: 'application/json',
            'Content-Type': 'application/json',
            ...headers,
        };
        const executor = async ({ document, variables }) => {
            const fetchResult = await fetch(pointer, {
                method,
                ...(method === 'POST'
                    ? {
                        body: JSON.stringify({ query: graphql.print(document), variables }),
                    }
                    : {}),
                headers: extraHeaders,
            });
            return fetchResult.json();
        };
        const clientSchema = await wrap.introspectSchema(executor);
        const remoteExecutableSchema = wrap.makeRemoteExecutableSchema({
            schema: utils.printSchemaWithDirectives(clientSchema, options),
            executor,
        });
        return {
            location: pointer,
            schema: remoteExecutableSchema,
        };
    }
    loadSync() {
        throw new Error('Loader Url has no sync mode');
    }
}

exports.UrlLoader = UrlLoader;
